packages/eslint-plugin-react-hooks/src/RulesOfHooks.ts

```diff
@@ -130,26 +128,26 @@ export default {
       url: 'https://reactjs.org/docs/hooks-rules.html',
     },
   },
-  create(context) {
-    let lastEffect = null;
-    const codePathReactHooksMapStack = [];
-    const codePathSegmentStack = [];
+  create(context: Rule.RuleContext) {
+    let lastEffect: CallExpression | null = null;
+    const codePathReactHooksMapStack: Map<Rule.CodePathSegment, Node[]>[] = [];
+    const codePathSegmentStack: Rule.CodePathSegment[] = [];
     const useEffectEventFunctions = new WeakSet();
 
     // For a given scope, iterate through the references and add all useEffectEvent definitions. We can
     // do this in non-Program nodes because we can rely on the assumption that useEffectEvent functions
     // can only be declared within a component or hook at its top level.
-    function recordAllUseEffectEventFunctions(scope) {
+    function recordAllUseEffectEventFunctions(scope: Scope.Scope): void {
       for (const reference of scope.references) {
         const parent = reference.identifier.parent;
         if (
-          parent.type === 'VariableDeclarator' &&
+          parent?.type === 'VariableDeclarator' &&
           parent.init &&
           parent.init.type === 'CallExpression' &&
           parent.init.callee &&
           isUseEffectEventIdentifier(parent.init.callee)
         ) {
-          for (const ref of reference.resolved.references) {
+          for (const ref of reference.resolved?.references || []) {
```

Comment:
For these types of changes, rather than relax the code by handling undefined behavior, I think we should prefer to throw/invariant instead if the constraint isn't being met. That helps us find any issues with our assumptions earlier instead of hiding it.

Reply:
Sounds good.  I'll add this before the the for loop
```ts
if (reference.resolved === null) {
  throw new Error('Unexpected undefined reference.resolved');
}
```